envision/server.py

```diff
@@ -54,30 +64,214 @@ def options(self):
         self.finish()
 
 
+class Frame:
+    def __init__(self, data: str, timestamp: int, next_=None):
+        # Time since epoch in seconds
+        self._timestamp = timestamp
+        self._data = data
+        self._size = sys.getsizeof(data)
+        self.next_ = next_
+
+    @property
+    def timestamp(self):
+        return self._timestamp
+
+    @property
+    def data(self):
+        return self._data
+
+    @data.setter
+    def data(self, data: str):
+        self._data = data
+        self._size = sys.getsizeof(data)
+
+    @property
+    def size(self):
+        return self._size
+
+
+class Frames:
+    def __init__(self, max_capacity_mb=500):
+        self._origin_timestamp = None
+        self._max_capacity = max_capacity_mb
+
+        # XXX: Index of timestamp in `self._timestamps` matches the index of the
+        # frame with the same timestamp in `self._frames`.
+        self._timestamps = []
+        self._frames = []
+
+    @property
+    def start_frame(self):
+        if len(self._frames) == 0:
+            return None
+
+        return self._frames[0]
+
+    @property
+    def start_time(self):
+        return self._origin_timestamp
+
+    @property
+    def elapsed_time(self):
+        return self._frames[-1].timestamp - self._origin_timestamp
+
+    def append(self, frame: Frame):
+        if len(self._frames) == 0:
+            self._origin_timestamp = frame.timestamp
+
+        self._enforce_max_capacity()
+
+        if len(self._frames) >= 1:
+            self._frames[-1].next_ = frame
+
+        self._frames.append(frame)
+        self._timestamps.append(frame.timestamp)
+
+    def __call__(self, timestamp):
+        """Finds the nearest frame according to the given timestamp."""
+        frame_idx = bisect.bisect_left(self._timestamps, timestamp)
+        if frame_idx >= len(self._frames):
+            frame_idx = -1
+
+        return self._frames[frame_idx]
+
+    def _enforce_max_capacity(self):
+        """Sample random frames and clear their data to ensure we're under the max
+        capacity size.
+        """
+        sizes = [frame.size for frame in self._frames]
+        while sum(sizes) * 1e-6 > self._max_capacity:
+            # XXX: randint(1, ...), we skip the start frame because that is a "safe"
+            #      frame we can always rely on being available.
+            idx_to_delete = random.randint(1, len(self._frames) - 1)
+            sizes[idx_to_delete] = 0
+            self._frames[idx_to_delete].data = None
```

Comment:
This frame list will still grow unbounded, how about we just remove the frame entirely. 
```suggestion
            if idx_to_delete > 0:
                self._frames[idx_to_delete - 1].next_ = self._frames[idx_to_delete].next_
            del self._frames[idx_to_delete]
```

Reply:
Yes, unbounded; we're talking about 2 ints and a reference: 28+28+8 = 64 bytes. At 50 FPS we're looking at 3200 bytes/sec ~275MB every 24 hours (may be a bit larger in practice).

Not great.

The reason for this design was to simplify seeking via the timestamps; but let me take another look.